### ISSUE TITLE
Validation loss demasiado alta

### ISSUE THREAD

**Issue by Ze6000** [miguelriemoliveira/savi_23-24#14]
Boa tarde professor @miguelriemoliveira,
Após testar com vários batch sizes, learning rates e diferentes modelos não consigo chegar a resultados favoráveis em termos de loss.
A train_loss é consistente, começa em valores perto de 1 a 2 e vai descendo até perto de zero, mas a validation_loss começa em 15 e apenas aumenta até aos 30 e... Deixo aqui uma imagem para ilustrar isso.

![Imagem WhatsApp 2024-01-29 às 16 50 39_10948d8e](https://github.com/miguelriemoliveira/savi_23-24/assets/88047954/45546521-c293-44c4-8784-7c7970292edb)

**Comment by Ze6000** [comment]
As imagens de treino e validação são as mesmas, tanto que dividi o dataset de várias maneiras diferentes para poder garantir que as 51 categorias estão em ambos.
O batch size já testei com 50, 100, 150, 200 e 300.
Tambem tentei com a amostra completa, mais de 120k de imagens, como com apenas 20k. Os resultados são semelhantes em qualquer uma destas combinações de batch size com total de imagens

**Comment by miguelriemoliveira** [comment]
Então deve ser bug no código. O código só com duas classes que dei na aula funciona? Quais são as diferenças?

**Comment by Ze6000** [comment]
O código de duas classes funciona bem e eu apenas fui adaptando para o transformar de binário para multi-class
No aspeto do treino ambos os códigos são iguais, não alterei nada aí. O que alterei foram as camadas que utilizo para construir os modelos, onde agora uso 3 fc para chegar ás 51 categorias, tambem alterei a forma como é atribuido label ás imagens em vez de if elif que tinhamos para uma função que vai guardando novos labels á medida que aparece uma imagem com um nome que o sistema ainda não conhece

**Comment by miguelriemoliveira** [comment]
É muito estranho. Corra uma experiência mais simples. Por exemplo só com 3 classes e 100 imagens. A coisa corre bem?

**Comment by Ze6000** [comment]
Todas estas são com apenas 3 categorias
Aqui tentei com 1000 treino, 100 validação e batchsize 500

![Captura de ecrã de 2024-01-30 16-27-36](https://github.com/miguelriemoliveira/savi_23-24/assets/88047954/b729f439-5b49-4d50-9dbe-844007187c77)

Aqui tentei com 10000 treino, 1000 validação e batchsize 500
![Captura de ecrã de 2024-01-30 16-32-31](https://github.com/miguelriemoliveira/savi_23-24/assets/88047954/aa61cf0b-4d5d-455e-8edf-a89d2d5731d6)

Aqui tentei com 100 treino, 10 validação e batchsize 500 
![Captura de ecrã de 2024-01-30 16-34-12](https://github.com/miguelriemoliveira/savi_23-24/assets/88047954/d40c54ec-a6fb-4a84-ad1b-2e0803a21d18)

Aqui tentei com 100 treino, 10 validação e batchsize 50 
![Captura de ecrã de 2024-01-30 16-36-09](https://github.com/miguelriemoliveira/savi_23-24/assets/88047954/dbbb7441-a764-49a7-8ef1-ab79e5789fcb)


Não sei bem que tipo de conclusões devo tirar disto, além de que o batchsize parece que deve ser quase tão grande quanto o numero de imagens. O que não faz muito sentido

**Comment by miguelriemoliveira** [comment]
Olá @Ze6000 ,

os resultados que mostra até parecem normais. Por exemplo no último caso, a curva de treino desce sempre, a de validação desce até certo ponto e depois começa a subir quando a rede começa a fazer overfitting aos dados.

Portanto, a única curva que me parece estranha é a de "Aqui tentei com 10000 treino, 1000 validação e batchsize 500".
As outras parecem normais.

Quanto ao erro se o batch for demasiado grande, é normal, está a esgotar a memória da GPU.

**Comment by Ze6000** [comment]
Eu tenho andado a pesquisar e não consigo perceber por que é que quando meto mais imagens a validation não funciona tão bem. Mas depois se tentar treinar modelos com menos imagens e por isso as losses um pouco melhores, chego ás métricas e os resultados não são bons. Com elas percebo que o programa raramente acerta o que está na imagem, originando estes resultados:
Precision = 0.007189542483660131
Recall = 0.017973856209150325
F1 score = 0.01027077497665733
Acho que mais vale focar no resto das tarefas e voltar a este problema mais tarde

**Comment by miguelriemoliveira** [comment]
Olá @Ze6000 ,

realmente, com esses valores de Precision e Recall diria que há um bug no código.

Cumprimentos,
Miguel
